File: Game.Core/BattleSimulation/Battler.cs
using Game.Core.Entities;
using static Game.Core.EAttribute;

namespace Game.Core.BattleSimulation
{
    /// <summary>
    /// The base class used to encapsulate a character for battle simulation.
    /// </summary>
    public abstract class Battler
    {
        /// <inheritdoc cref="BattleAttributes" />
        public abstract BattleAttributes Attributes { get; set; }

        /// <summary>
        /// The health for the battler for the current simulation.
        /// </summary>
        public abstract double CurrentHealth { get; set; }

        /// <summary>
        /// The skills that the battler will use during simulation.
        /// </summary>
        public abstract List<BattleSkill> Skills { get; set; }

        /// <summary>
        /// The current level of the battler.
        /// </summary>
        public abstract int Level { get; set; }

        /// <summary>
        /// True when the battler has been defeated in the simulation.
        /// </summary>
        public bool IsDead => CurrentHealth <= 0;

        /// <summary>
        /// Increases each <see cref="BattleSkill.ChargeTime"/> by the given <paramref name="timeDelta"/>. If any go over the <see cref="Skill.CooldownMs"/>
        /// then the <see cref="BattleSkill.ChargeTime"/> will be reset to 0 and the <see cref="BattleSkill"/> returned.
        /// </summary>
        /// <param name="timeDelta">The amount in milliseconds to advance each <see cref="BattleSkill.ChargeTime"/>.</param>
        /// <returns>A <see cref="List{T}"/> of each <see cref="BattleSkill"/> where the <see cref="BattleSkill.ChargeTime"/> reached the <see cref="Skill.CooldownMs"/>.</returns>
        public List<BattleSkill> AdvancedCooldowns(int timeDelta)
        {
            var firedSkills = new List<BattleSkill>();
            var cdMultiplier = 1 + (Attributes[CooldownRecovery] / 100);
            foreach (var skill in Skills)
            {
                skill.ChargeTime += timeDelta * cdMultiplier;
                if (skill.ChargeTime > skill.Data.CooldownMs)
                {
                    firedSkills.Add(skill);
                    skill.ChargeTime = 0;
                }
            }

            return firedSkills;
        }

        /// <summary>
        /// Decreases the <see cref="CurrentHealth"/> by the <paramref name="rawDamage"/> after reduction using the <see cref="Defense"/> of the <see cref="Battler"/>.
        /// </summary>
        /// <param name="rawDamage">The damage to apply before reductions.</param>
        public void TakeDamage(double rawDamage)
        {
            var damage = rawDamage - Attributes[Defense];
            damage = damage > 0 ? damage : 0;
            CurrentHealth -= damage;
        }
    }
}

File: UI/new-svelte/src/lib/battle/battler.ts
import { Skill } from './skill';
import { BattleAttributes } from './battle-attributes';
import { IBattlerAttribute, EAttribute } from '$lib/api';
import { staticData } from '$stores';

interface BattlerData {
	attributes: IBattlerAttribute[];
	selectedSkills: number[];
	level: number;
	name: string;
}

const maxSkills = 4;
let battlerId = 0;

export class Battler {
	public id = battlerId++;
	public name = '';
	public level = 0;
	public currentHealth = 0;
	public attributes: BattleAttributes = new BattleAttributes();
	public skills: (Skill | undefined)[] = [];
	public maxSkills: typeof maxSkills = maxSkills;
	public cdMultiplier = 1;
	public isDead = true;

	constructor(battlerData?: BattlerData, additionalAtttributes?: IBattlerAttribute[]) {
		this.reset(battlerData, additionalAtttributes);
	}

	public advanceCooldowns(timeDelta: number) {
		const firedSkills: Skill[] = [];
		for (const skill of this.skills) {
			if (skill) {
				skill.chargeTime += timeDelta * this.cdMultiplier;
				if (skill.chargeTime >= skill.cooldownMs) {
					firedSkills.push(skill);
					skill.chargeTime = 0;
				}
			}
		}
		return firedSkills;
	}

	public updateRenderCooldowns(renderDelta: number) {
		for (const skill of this.skills) {
			if (skill) {
				skill.renderChargeTime = Math.min(
					skill.chargeTime + renderDelta * this.cdMultiplier,
					skill.cooldownMs
				);
			}
		}
	}

	public takeDamage(rawDamage: number) {
		let damage = rawDamage - this.attributes.getValue(EAttribute.Defense);
		if (damage < 0) {
			damage = 0;
		}
		this.currentHealth -= damage;
		this.isDead = this.currentHealth <= 0;
		return damage;
	}

	public reset(battlerData?: BattlerData, additionalAtttributes?: IBattlerAttribute[]) {
		if (battlerData) {
			const atts = additionalAtttributes
				? [...battlerData.attributes, ...additionalAtttributes]
				: battlerData.attributes;

			this.attributes.setData(atts);
			this.level = battlerData.level;
			this.name = battlerData.name;
			this.skills = this.fillSelectedSkills(battlerData);
		}

		this.currentHealth = this.attributes.getValue(EAttribute.MaxHealth);
		this.cdMultiplier = 1 + this.attributes.getValue(EAttribute.CooldownRecovery) / 100;
		this.isDead = false;
	}

	private fillSelectedSkills(battlerData: BattlerData) {
		const skillData = staticData.skills;
		const skills: (Skill | undefined)[] = battlerData.selectedSkills.map(
			(skillId) => new Skill(skillData[skillId], this)
		);
		while (skills.length < maxSkills) {
			skills.push(undefined);
		}
		return skills;
	}
}
